driver-core/src/main/com/mongodb/operation/ChangeStreamBatchCursor.java

```diff
@@ -0,0 +1,164 @@
+                return convertResults(queryBatchCursor.next());
+            }
+        });
+    }
+
+    @Override
+    public List<T> tryNext() {
+        return resumeAbleOperation(new Function<QueryBatchCursor<RawBsonDocument>, List<T>>() {
+            @Override
+            public List<T> apply(final QueryBatchCursor<RawBsonDocument> queryBatchCursor) {
+                return convertResults(queryBatchCursor.tryNext());
+            }
+        });
+    }
+
+    @Override
+    public void close() {
+        wrapped.close();
+    }
+
+    @Override
+    public void setBatchSize(final int batchSize) {
+        wrapped.setBatchSize(batchSize);
+    }
+
+    @Override
+    public int getBatchSize() {
+        return wrapped.getBatchSize();
+    }
+
+    @Override
+    public ServerCursor getServerCursor() {
+        return wrapped.getServerCursor();
+    }
+
+    @Override
+    public ServerAddress getServerAddress() {
+        return wrapped.getServerAddress();
+    }
+
+    @Override
+    public void remove() {
+        throw new UnsupportedOperationException("Not implemented yet!");
+    }
+
+    private List<T> convertResults(final List<RawBsonDocument> rawDocuments) {
+        List<T> results = null;
+        if (rawDocuments != null) {
+            results = new ArrayList<T>();
+            for (RawBsonDocument rawDocument : rawDocuments) {
+                resumeToken = rawDocument.getDocument("_id");
+                results.add(rawDocument.decode(changeStreamOperation.getCodec()));
+            }
+        }
+        return results;
+    }
+
+    <R> R resumeAbleOperation(final Function<QueryBatchCursor<RawBsonDocument>, R> function) {
+        if (changeStreamOperation.isResumable()) {
+            try {
+                return function.apply(wrapped);
+            } catch (Exception e) {
+                if (e instanceof MongoServerException && !(e instanceof MongoNotPrimaryException || e instanceof MongoQueryException)) {
+                    throw (MongoServerException) e;
+                }
+
+                killPreviousCursor(wrapped.getServerCursor());
+                wrapped = ((ChangeStreamBatchCursor<T>) new ChangeStreamOperation<T>(changeStreamOperation, resumeToken)
+                        .execute(binding)).getWrapped();
+                return function.apply(wrapped);
+            }
+        } else {
+            return function.apply(wrapped);
+        }
+    }
+
+    private void killPreviousCursor(final ServerCursor serverCursor) {
+        if (serverCursor != null) {
+            MongoNamespace namespace = changeStreamOperation.getNamespace();
+            BsonDocument command = new BsonDocument("killCursors", new BsonString(namespace.getCollectionName()))
+                    .append("cursors", new BsonArray(singletonList(new BsonInt64(serverCursor.getId()))));
+            try {
+                executeWrappedCommandProtocol(binding, namespace.getDatabaseName(), command, BSON_DOCUMENT_CODEC);
```

Comment:
This isn't quite right.  Since all you have is the ReadBinding, the killCursors command will be executed against any server in the cluster that satisfies the read preference.  But it only makes sense to send it to the same server as the cursor is on.

I think this can be fixed by instead just calling wrapped.close()